wolfcrypt/src/random.c

```diff
@@ -800,6 +809,779 @@ int wc_RNG_TestSeed(const byte* seed, word32 seedSz)
+ *
+ * @param [out] raw  Buffer to hold raw entropy data.
+ * @param [in]  cnt  Number of bytes of raw entropy to get.
+ * @return  0 on success.
+ * @return  Negative when creating a thread fails - when no high resolution
+ * clock available.
+ */
+int wc_Entropy_GetRawEntropy(unsigned char* raw, int cnt)
+{
+    int ret = 0;
+
+#ifdef ENTROPY_MEMUSE_THREADED
+    /* Start the counter thread as a proxy for time counter. */
+    ret = Entropy_StartThread();
+    if (ret == 0)
+#endif
+    {
+        Entropy_GetNoise(raw, cnt);
+    }
+#ifdef ENTROPY_MEMUSE_THREADED
+    /* Stop the counter thread to avoid thrashing the system. */
+    Entropy_StopThread();
+#endif
+
+    return ret;
+}
+
+#if ENTROPY_MIN == 1
+/* SP800-90b 4.4.1 - Repetition Test
+ * C = 1 + upper(-log2(alpha) / H)
+ * When alpha = 2^-30 and H = 1,
+ * C = 1 + upper(30 / 1) = 31
+ */
+#define REP_CUTOFF           31
+#else
+#error "Minimim entropy not defined to a recongnized value."
+#endif
+
+/* Have valid previous sample for repetition test. */
+static int rep_have_prev = 0;
+/* Previous sample value. */
+static byte rep_prev_noise;
+
+static void Entropy_HealthTest_Repetition_Reset()
+{
+    /* No previous stored. */
+    rep_have_prev = 0;
+    /* Clear previous. */
+    rep_prev_noise = 0;
+}
+
+/* Test sample value with repetition test.
+ *
+ * @param [in] noise  Sample to test.
+ * @return  0 on success.
+ * @return  ENTROPY_RT_E on failure.
+ */
+static int Entropy_HealthTest_Repetition(byte noise)
+{
+    int ret = 0;
+    /* Number of times previous value has been seen continuously. */
+    static int rep_cnt = 0;
+
+    /* If we don't have a previous then store this one for next time. */
+    if (!rep_have_prev) {
+        rep_prev_noise = noise;
+        rep_have_prev = 1;
+    }
+    /* Check whether this sample matches last. */
+    else if (noise == rep_prev_noise) {
+        /* Update count of repititions. */
+        rep_cnt++;
+        /* Fail if we reach cutoff. */
+        if (rep_cnt >= REP_CUTOFF) {
+        #ifdef WOLFSSL_DEBUG_ENTROPY_MEMUSE
+            fprintf(stderr, "REPETITION FAILED: %d\n", noise);
+        #endif
+            Entropy_HealthTest_Repetition_Reset();
+            ret = ENTROPY_RT_E;
+        }
+    }
+    else {
+        /* Cache new previous and seen one so far. */
+        rep_prev_noise = noise;
+        rep_cnt = 1;
+    }
+
+    return ret;
+}
+
+/* SP800-90b 4.4.2 - Adaptive Proportion Test
+ * Para 2
+ *   ... The window size W is selected based on the alphabet size ... 512 if
+ *   the noise source is not binary ...
+ */
+#define PROP_WINDOW_SIZE     512
+#if ENTROPY_MIN == 1
+/* SP800-90b 4.4.2 - Adaptive Proportion Test
+ * Note 10
+ * C = 1 + CRITBINOM(W, power(2,( -H)),1-alpha)
+ * alpa = 2^-30 = POWER(2,-30), H = 1, W = 512
+ * C = 1 + CRITBINOM(512, 0.5, 1-POWER(2,-30)) = 1 + 324 = 325
+ */
+#define PROP_CUTOFF          325
+#else
+#error "Minimim entropy not defined to a recongnized value."
```

Comment:
Spelling: `#error "Minimum entropy not defined to a recognized value."`

Reply:
fixed